### PR TITLE
Treat duplicative certificate requests as renewals by default

### LINKED ISSUE #411: Sensible default handling of lineages when the client is re-run.
Suppose the following events occur:

1. On Tuesday the user runs <tt>letsencyrpt</tt> on all her domains (perhaps she has 5)
2. On Wednesday she adds a new vhost
3. On Thursday she runs <tt>letsencrypt</tt> on all her names again (now she has 6)

Step 3 above shouldn't create a second certificate lineage which is renewed along with the first; by default it should create a successor to the first.  There can be a command line flag (perhaps <tt>--duplicate</tt>) to actually create a second, if that's what you really want.

Now suppose this case:

1. On Tuesday the user runs <tt>letsencrypt --vhost sni</tt> to make 5 certs
2. On Wednesday she adds a new vhost
3. On Thursday she runs <tt>letsencrypt --vhost sni</tt> again to make 6 certs

The second command should either not make new certs for the 5 existing ones, or should make successors in their lineages, unless <tt>--duplicate</tt> is present.

### PR DESCRIPTION
**Pull request by schoen**
This worked properly in several configurations and permutations for me, and should fix #411.

I have also added a version of the `--duplicate` option to override this behavior. Following a previous discussion, this has to be specified manually because we concluded that most users are very unlikely to actually want duplicate certificates.

Coverage went down and this PR currently fails coverage on Travis by about 1%. There are no existing tests for the portions of `client.py` and `cli.py` that the PR modifies so I don't have anything to build on.

### CODE REVIEW

**Comment by kuba** [comment]
Please split out new functionality from `run` into a separate function. This will make the code not only more readable, but also modular/composable, thus allowing you to write the necessary unittests :)

**Comment by schoen** [comment]
@kuba, I've become convinced of that and already started trying to split it up a bit! I think that's the best course.

**Comment by schoen** [comment]
I disabled the `duplicate-code` check in pylint because I deliberately have duplicate code in a test that I copied from another test, and I couldn't figure out how to disable the duplicate code check for only that part of the code. (The normal pylint disable comment mechanism didn't seem to work for this.)

File: .pylintrc [hunk 1]
```diff
@@ -38,7 +38,7 @@ load-plugins=linter_plugin
 # --enable=similarities". If you want to run only the classes checker, but have
 # no Warning level messages displayed, use"--disable=all --enable=classes
 # --disable=W"
-disable=fixme,locally-disabled,abstract-class-not-used
+disable=fixme,locally-disabled,abstract-class-not-used,duplicate-code
```
Comment (kuba) [inline .pylintrc]: Could you refactor the code instead of disabling this warning? Unfortunately, I don't think it can be disabled globally, so it has a big impact on the rest of the project.

File: letsencrypt/cli.py [hunk 2]
```diff
@@ -1,7 +1,9 @@
 """Let's Encrypt CLI."""
 # TODO: Sanity check all input.  Be sure to avoid shell code etc...
+
```
Comment (kuba) [inline letsencrypt/cli.py]: should be removed

File: letsencrypt/cli.py [hunk 3]
```diff
@@ -1,7 +1,9 @@
...
 import argparse
 import atexit
+import configobj
```
Comment (kuba) [inline letsencrypt/cli.py]: should be imported lower down, below `configargparse`. this section is for stdlib imports

File: letsencrypt/tests/cli_test.py [hunk 4]
```diff
@@ -162,5 +167,73 @@ def test_no_accounts_email(self):
         self.assertEqual("other email", self.args.email)
 
 
+class DuplicativeCertsTest(unittest.TestCase):
+
+    def setUp(self):
+        # The stuff below is taken from renewer_test.py
```
Comment (kuba) [inline letsencrypt/tests/cli_test.py]: well, just export it in renewer_test and import here instead of copying

File: letsencrypt/cli.py [hunk 5]
```diff
@@ -159,7 +163,34 @@ def _init_le_client(args, config, authenticator, installer):
...
+
+    configs_dir = renew_config.renewal_configs_dir
+    for renewal_file in os.listdir(configs_dir):
+        full_path = os.path.join(configs_dir, renewal_file)
+        rc_config = configobj.ConfigObj(renew_config.renewer_config_file)
+        rc_config.merge(configobj.ConfigObj(full_path))
+        rc_config.filename = full_path
+        cli_config = configuration.RenewerConfiguration(config)
+
+        candidate_lineage = storage.RenewableCert(rc_config, None, cli_config)
```
Comment (kuba) [inline letsencrypt/cli.py]: Please use explicit keywords in the call, i.e. `foo('bar', baz=None)` instead of `foo('bar', None)`. Otherwise it's necessary to go to the `storage.py` and look it up, which is unnecessarily tedious. readability--

File: letsencrypt/cli.py [hunk 6]
```diff
@@ -159,7 +163,34 @@ def _init_le_client(args, config, authenticator, installer):
...
+            identical_names_cert = (renewal_file, candidate_lineage)
+        elif candidate_names.issubset(set(domains)):
+            subset_names_cert = (renewal_file, candidate_lineage,
+                                 candidate_lineage.names())
+
+    return identical_names_cert, subset_names_cert
+
+
+def run(args, config, plugins):  # pylint: disable=too-many-branches,too-many-locals
+
```
Comment (kuba) [inline letsencrypt/cli.py]: whitespace should be removed

File: letsencrypt/cli.py [hunk 7]
```diff
@@ -159,7 +163,34 @@ def _init_le_client(args, config, authenticator, installer):
...
+        # TODO: Handle these differently depending on whether they are
+        #       expired or still valid?
+        candidate_names = set(candidate_lineage.names())
+        if candidate_names == set(domains):
+            identical_names_cert = (renewal_file, candidate_lineage)
+        elif candidate_names.issubset(set(domains)):
+            subset_names_cert = (renewal_file, candidate_lineage,
+                                 candidate_lineage.names())
+
+    return identical_names_cert, subset_names_cert
```
Comment (kuba) [inline letsencrypt/cli.py]: Something is wrong here... This function seems to return data only for the last `renewal_file` in `os.listdir(configs_dir)`, because you don't accumulate results in `identical_names_cert` and `subset_names_cert` for each loop iteration, but overwrite it instead.

File: letsencrypt/cli.py [hunk 8]
```diff
@@ -159,7 +163,34 @@ def _init_le_client(args, config, authenticator, installer):
...
+    """Find existing certs that duplicate the request."""
+
+    identical_names_cert, subset_names_cert = None, None
+
+    configs_dir = renew_config.renewal_configs_dir
+    for renewal_file in os.listdir(configs_dir):
+        full_path = os.path.join(configs_dir, renewal_file)
+        rc_config = configobj.ConfigObj(renew_config.renewer_config_file)
+        rc_config.merge(configobj.ConfigObj(full_path))
+        rc_config.filename = full_path
```
Comment (kuba) [inline letsencrypt/cli.py]: code for initializing renewer configuration seems to follow the same pattern as in other parts of the source code. refactor?

```python
def load_renewer_config(renew_config, path):
    rc_config = configobj.ConfigObj(renew_config.renewer_config_file)
    rc_config.merge(configobj.ConfigObj(full_path))
    rc_config.filename = full_path
    return rc_config
```

File: letsencrypt/cli.py [hunk 9]
```diff
@@ -159,7 +163,34 @@ def _init_le_client(args, config, authenticator, installer):
...
+        cli_config = configuration.RenewerConfiguration(config)
```
Comment (kuba) [inline letsencrypt/cli.py]: unnecessarily initialized in each loop iteration, move above `for renewal_file ...`

File: letsencrypt/cli.py (hunk 8 above)
Comment (kuba) [inline letsencrypt/cli.py]: In fact, in https://github.com/letsencrypt/letsencrypt/blob/master/letsencrypt/renewer.py#L153-L157, it's marked as `# TODO: this is a dirty hack!`.

File: letsencrypt/cli.py (hunk 5 above)
Comment (kuba) [inline letsencrypt/cli.py]: also, this can throw an error if config file is broken and it's not caught anywhere

File: letsencrypt/cli.py [hunk 10]
```diff
@@ -181,15 +212,88 @@ def run(args, config, plugins):
...
+        else:
+            reporter_util = zope.component.getUtility(interfaces.IReporter)
+            reporter_util.add_message((
+                "To obtain a new certificate that {0} an existing certificate "
+                "in its domain-name coverage, you must use the --duplicate "
+                "option.\n\nFor example:\n\n{1} --duplicate {2}").format(
+                    "duplicates" if identical_names_cert else "overlaps with",
+                    sys.argv[0], " ".join(sys.argv[1:])),
+                                      reporter_util.HIGH_PRIORITY, True)
+            sys.exit(1)
```
Comment (kuba) [inline letsencrypt/cli.py]: please use `return 1` or `return "some message"` instead. rest is handled in https://github.com/letsencrypt/letsencrypt/blob/master/letsencrypt/cli.py#L788-L792

File: letsencrypt/cli.py [hunk 11]
```diff
@@ -483,6 +587,9 @@ def create_parser(plugins, args):
     #for subparser in parser_run, parser_auth, parser_install:
     #    subparser.add_argument("domains", nargs="*", metavar="domain")
     helpful.add(None, "-d", "--domains", metavar="DOMAIN", action="append")
+    helpful.add(
+        None, "-D", "--duplicate", dest="duplicate", action="store_true",
```
Comment (kuba) [inline letsencrypt/cli.py]: do we need a short flag as well?

File: letsencrypt/client.py [hunk 12]
```diff
@@ -378,8 +378,6 @@ def deploy_certificate(self, domains, privkey_path, cert_path, chain_path):
         # sites may have been enabled / final cleanup
         self.installer.restart()
 
-        display_ops.success_installation(domains)
```
Comment (kuba) [inline letsencrypt/client.py]: why remove?

File: letsencrypt/storage.py [hunk 13]
```diff
@@ -421,6 +422,24 @@ def notafter(self, version=None):
...
+        :rtype: `list` of `str`
+
+        """
+        if version is None:
+            target = self.current_target("cert")
+        else:
+            target = self.version("cert", version)
+        with open(target) as f:
+            sans = crypto_util.get_sans_from_cert(f.read())
+        return sans
```
Comment (kuba) [inline letsencrypt/storage.py]:
```pytohn
with open(target) as f:
    return crypto_util.get_sans_from_cert(f.read()
```

File: letsencrypt/tests/cli_test.py [hunk 14]
```diff
@@ -1,4 +1,5 @@
 """Tests for letsencrypt.cli."""
+import configobj
```
Comment (kuba) [inline letsencrypt/tests/cli_test.py]: import in wrong section

File: letsencrypt/cli.py [hunk 15]
```diff
@@ -181,15 +212,88 @@ def run(args, config, plugins):
...
+    # related to an existing certificate.  (config.duplicate, which
+    # is set with --duplicate, skips all of this logic and forces any
+    # kind of certificate to be obtained with treat_as_renewal = False.)
+    if not config.duplicate:
+        identical_names_cert, subset_names_cert = _find_duplicative_certs(
+            domains, config, configuration.RenewerConfiguration(config))
+        # I am not sure whether that correctly reads the systemwide
+        # configuration file.
+        question = None
+        if identical_names_cert:
```
Comment (kuba) [inline letsencrypt/cli.py]: `if identical_names cert is not None`

File: letsencrypt/cli.py [hunk 16]
```diff
@@ -181,15 +212,88 @@ def run(args, config, plugins):
...
+            question = (
+                "You have an existing certificate that contains exactly the "
+                "same domains you requested (ref: {0})\n\nDo you want to "
+                "renew and replace this certificate with a newly-issued one?"
+                ).format(identical_names_cert[0])
+        elif subset_names_cert:
```
Comment (kuba) [inline letsencrypt/cli.py]: `elif susbset_names_cert is not None:`

File: letsencrypt/cli.py [hunk 17]
```diff
@@ -159,7 +163,34 @@ def _init_le_client(args, config, authenticator, installer):
...
+
+        candidate_lineage = storage.RenewableCert(rc_config, None, cli_config)
+        # TODO: Handle these differently depending on whether they are
+        #       expired or still valid?
+        candidate_names = set(candidate_lineage.names())
+        if candidate_names == set(domains):
+            identical_names_cert = (renewal_file, candidate_lineage)
+        elif candidate_names.issubset(set(domains)):
+            subset_names_cert = (renewal_file, candidate_lineage,
+                                 candidate_lineage.names())
```
Comment (kuba) [inline letsencrypt/cli.py]: could you use `collections.namedtuple` for the return value?, `subset_names_cert[x]` used later in the code is not very readable

File: letsencrypt/cli.py [hunk 18]
```diff
@@ -181,15 +212,88 @@ def run(args, config, plugins):
...
+                question, "Replace", "Cancel"):
+            treat_as_renewal = True
+        else:
+            reporter_util = zope.component.getUtility(interfaces.IReporter)
+            reporter_util.add_message((
+                "To obtain a new certificate that {0} an existing certificate "
+                "in its domain-name coverage, you must use the --duplicate "
+                "option.\n\nFor example:\n\n{1} --duplicate {2}").format(
+                    "duplicates" if identical_names_cert else "overlaps with",
+                    sys.argv[0], " ".join(sys.argv[1:])),
```
Comment (kuba) [inline letsencrypt/cli.py]: please use `args` instead of `sys.argv`

File: letsencrypt/cli.py [hunk 19]
```diff
@@ -181,15 +212,88 @@ def run(args, config, plugins):
...
+                                      reporter_util.HIGH_PRIORITY, True)
```
Comment (kuba) [inline letsencrypt/cli.py]: again, please use `foo=True` syntax, as it's not clear what `True` refers to

File: letsencrypt/tests/cli_test.py [hunk 20]
```diff
@@ -162,5 +167,73 @@ def test_no_accounts_email(self):
...
+            with open(where, "w") as f:
+                f.write(kind)
+
+        # Here we will use test_rc to create duplicative stuff
+
+    def tearDown(self):
+        shutil.rmtree(self.tempdir)
+
+    def test_find_duplicative_names(self):
+        from letsencrypt.cli import _find_duplicative_certs
```
Comment (kuba) [inline letsencrypt/tests/cli_test.py]: i'm surprised you don't have to `pylint: disable=protected-access` here; what's going on?

File: letsencrypt/cli.py [hunk 21]
```diff
@@ -181,15 +212,88 @@ def run(args, config, plugins):
...
     le_client = _init_le_client(args, config, authenticator, installer)
-    lineage = le_client.obtain_and_enroll_certificate(
-        domains, authenticator, installer, plugins)
-    if not lineage:
-        return "Certificate could not be obtained"
-    le_client.deploy_certificate(
-        domains, lineage.privkey, lineage.cert, lineage.chain)
-    le_client.enhance_config(domains, args.redirect)
+    if treat_as_renewal:
+        if identical_names_cert:
```
Comment (kuba) [inline letsencrypt/cli.py]: `if identical_names_cert is not None`

File: letsencrypt/client.py (hunk 12 above)
Comment (kuba) [inline letsencrypt/client.py]: Ok, I can see it's been moved to `cli.py`. I'm actually very happy about this change, as it allows us to decouple `IDisplay` from the API (`client.py`)

File: letsencrypt/tests/cli_test.py [hunk 22]
```diff
@@ -11,6 +12,10 @@
 from letsencrypt import account
 from letsencrypt import configuration
 from letsencrypt import errors
+from letsencrypt import storage
+
+from letsencrypt.storage import ALL_FOUR
```
Comment (kuba) [inline letsencrypt/tests/cli_test.py]: please do `from letsencrypt import storage; storage.ALL_FOUR`

**Comment by kuba** [comment]
Run `auth` twice without any special flags (without `--duplicate`):

```
# tree /etc/letsencrypt/
/etc/letsencrypt/
|-- accounts
|   `-- acme-staging.api.letsencrypt.org
|       `-- acme
|           `-- new-reg
|               `-- be0a12cf6888767b3dbff4006fee7e15
|                   |-- meta.json
|                   |-- private_key.json
|                   `-- regr.json
|-- archive
|   |-- kuba.us.to
|   |   |-- cert1.pem
|   |   |-- chain1.pem
|   |   |-- fullchain1.pem
|   |   `-- privkey1.pem
|   `-- kuba.us.to-0001
|       |-- cert1.pem
|       |-- chain1.pem
|       |-- fullchain1.pem
|       `-- privkey1.pem
|-- certs
|   |-- 0000_csr-letsencrypt.pem
|   `-- 0001_csr-letsencrypt.pem
|-- configs
|   |-- kuba.us.to-0001.conf
|   `-- kuba.us.to.conf
|-- keys
|   |-- 0000_key-letsencrypt.pem
|   `-- 0001_key-letsencrypt.pem
`-- live
    |-- kuba.us.to
    |   |-- cert.pem -> ../../archive/kuba.us.to/cert1.pem
    |   |-- chain.pem -> ../../archive/kuba.us.to/chain1.pem
    |   |-- fullchain.pem -> ../../archive/kuba.us.to/fullchain1.pem
    |   `-- privkey.pem -> ../../archive/kuba.us.to/privkey1.pem
    `-- kuba.us.to-0001
        |-- cert.pem -> ../../archive/kuba.us.to-0001/cert1.pem
        |-- chain.pem -> ../../archive/kuba.us.to-0001/chain1.pem
        |-- fullchain.pem -> ../../archive/kuba.us.to-0001/fullchain1.pem
        `-- privkey.pem -> ../../archive/kuba.us.to-0001/privkey1.pem

14 directories, 25 files
```

After `rm -rf /etc/letsencrypt /var/lib/letsencrypt, run `auth` twice with --duplicate:

```
# tree /etc/letsencrypt/
/etc/letsencrypt/
|-- accounts
|   `-- acme-staging.api.letsencrypt.org
|       `-- acme
|           `-- new-reg
|               `-- 660cc7b792f5a5f5c4c4bc8179b2f9c9
|                   |-- meta.json
|                   |-- private_key.json
|                   `-- regr.json
|-- archive
|   |-- kuba.us.to
|   |   |-- cert1.pem
|   |   |-- chain1.pem
|   |   |-- fullchain1.pem
|   |   `-- privkey1.pem
|   `-- kuba.us.to-0001
|       |-- cert1.pem
|       |-- chain1.pem
|       |-- fullchain1.pem
|       `-- privkey1.pem
|-- certs
|   |-- 0000_csr-letsencrypt.pem
|   `-- 0001_csr-letsencrypt.pem
|-- configs
|   |-- kuba.us.to-0001.conf
|   `-- kuba.us.to.conf
|-- keys
|   |-- 0000_key-letsencrypt.pem
|   `-- 0001_key-letsencrypt.pem
`-- live
    |-- kuba.us.to
    |   |-- cert.pem -> ../../archive/kuba.us.to/cert1.pem
    |   |-- chain.pem -> ../../archive/kuba.us.to/chain1.pem
    |   |-- fullchain.pem -> ../../archive/kuba.us.to/fullchain1.pem
    |   `-- privkey.pem -> ../../archive/kuba.us.to/privkey1.pem
    `-- kuba.us.to-0001
        |-- cert.pem -> ../../archive/kuba.us.to-0001/cert1.pem
        |-- chain.pem -> ../../archive/kuba.us.to-0001/chain1.pem
        |-- fullchain.pem -> ../../archive/kuba.us.to-0001/fullchain1.pem
        `-- privkey.pem -> ../../archive/kuba.us.to-0001/privkey1.pem

14 directories, 25 files
```

I expected to have only `live/kuba.us.to` in one of the runs, but there is no difference. Also, I didn't get any confirmation questions about already detected domain.

How is that PR meant to work? Could you add integration tests?

File: letsencrypt/cli.py (hunk 6 above)
Comment (schoen) [inline letsencrypt/cli.py]: Which whitespace?

File: letsencrypt/cli.py (hunk 7 above)
Comment (schoen) [inline letsencrypt/cli.py]: That is deliberate. The client will only offer to replace one existing certificate, even if several alternatives exist. I think we thought that it would confuse or annoy the user to have to choose which one to replace, although I can agree that some users would want the option.

If you'd like to propose a different behavior, could you please open an issue to discuss it, or bring it up in #411? I'm happy to discuss what the behavior should be, I'm just pointing out that for the time being this is intentional.

File: letsencrypt/cli.py (hunk 8 above)
Comment (schoen) [inline letsencrypt/cli.py]: I'm still extremely confused about what the correct way to initialize the various configuration objects should be. I don't think we've solved this (in several places) and I'd encourage a "configuration object hunt" or "configuration object review". (I know what this one does and I think it's probably appropriate here, but I don't fully understand what some of the other ones do.)

File: letsencrypt/cli.py (hunk 9 above)
Comment (schoen) [inline letsencrypt/cli.py]: Yep! Thanks.

File: letsencrypt/cli.py (hunk 11 above)
Comment (schoen) [inline letsencrypt/cli.py]: Isn't -D the short flag? Or are you saying that maybe there shouldn't be a short flag?

File: letsencrypt/client.py (hunk 12 above)
Comment (schoen) [inline letsencrypt/client.py]: Great!

File: letsencrypt/cli.py (hunk 17 above)
Comment (schoen) [inline letsencrypt/cli.py]: I figured out a way to avoid having the return values be tuples, and just return the individual lineages.

File: letsencrypt/cli.py (hunk 18 above)
Comment (schoen) [inline letsencrypt/cli.py]: `TypeError: 'Namespace' object does not support indexing`

File: letsencrypt/cli.py (hunk 19 above)
Comment (schoen) [inline letsencrypt/cli.py]: I didn't know what it referred to because I copied this invocation from elsewhere. It turns out we don't need it here.

File: letsencrypt/tests/cli_test.py (hunk 20 above)
Comment (schoen) [inline letsencrypt/tests/cli_test.py]: I don't know.

**Comment by schoen** [comment]
@Kuba, I was going to get very confused, but I realized that the changes implemented in this PR currently only apply to `run` and not to `auth`. I think @pde was going to change the verbs around quite a bit so I'm not sure if it's worthwhile to try to do this in `auth` right now, but we should at least warn @pde about the issue.

File: letsencrypt/cli.py (hunk 6 above)
Comment (kuba) [inline letsencrypt/cli.py]: empty line between `def` and `"""`

File: letsencrypt/cli.py (hunk 11 above)
Comment (kuba) [inline letsencrypt/cli.py]: sorry, I meant that probably there shouldn't be a short flag

Comment (schoen) [inline letsencrypt/cli.py]: I'm happy to get rid of it; do I just replace the string with None?

Comment (schoen) [inline letsencrypt/cli.py]: Or just remove the string from the arguments list entirely?

Comment (kuba) [inline letsencrypt/cli.py]: just remove that particular string, i.e. `None` should not be used

File: letsencrypt/cli.py [hunk 23]
```diff
@@ -163,7 +166,34 @@ def _init_le_client(args, config, authenticator, installer):
...
+
+    configs_dir = renew_config.renewal_configs_dir
+    cli_config = configuration.RenewerConfiguration(config)
+    for renewal_file in os.listdir(configs_dir):
+        full_path = os.path.join(configs_dir, renewal_file)
+        rc_config = configobj.ConfigObj(renew_config.renewer_config_file)
+        rc_config.merge(configobj.ConfigObj(full_path))
+        rc_config.filename = full_path
+
+        candidate_lineage = storage.RenewableCert(rc_config, config_opts=None,
```
Comment (kuba) [inline letsencrypt/cli.py]: As mentioned before, this might throw exceptions. This is in particular important, because we are reading user input, a renewal configuration file that might be broken. Please at least add a `TODO` comment inline, so we fix before launch. Alternatively you could fix it here :) How about:

```python
try:
    candidate_lineage = storage.Renewa...
except SomeException:
    logger.warning("Renewal configuration file %s is broken. Skipping.", full_path)
    continue
```

File: letsencrypt/cli.py (hunk 8 above)
Comment (kuba) [inline letsencrypt/cli.py]: I could try to hunt it down, but don't want to interfere with your work and cause merge conflicts. Do you plan to work on the relevant code this weekend?

Comment (schoen) [inline letsencrypt/cli.py]: I don't plan to change anything that's in any way related to configuration objects. I will probably only make the other 2-3 changes that we've just discussed, none of which should conflict with this.

File: letsencrypt/cli.py (hunk 7 above)
Comment (kuba) [inline letsencrypt/cli.py]: I'm OK with this behavior (at least for now). However, could you please document it in the function docstring?

File: letsencrypt/cli.py (hunk 17 above)
Comment (kuba) [inline letsencrypt/cli.py]: great!

File: letsencrypt/cli.py (hunk 18 above)
Comment (kuba) [inline letsencrypt/cli.py]: Oh sorry, my bad! `main` accepts `cli_args` rather than hard coding `sys.argv` (important for tests) - so it should be probably passed down to the subcommand handler functions. I'll look into this later.

File: letsencrypt/cli.py (hunk 19 above)
Comment (kuba) [inline letsencrypt/cli.py]: :)

**Comment by kuba** [comment]
Hah, sorry for the confusion. I'll try it with `run` later this day. :)

Please ping me when you re-enable `duplicate-code` check so I can have another look.

**Comment by jdkasten** [comment]
I took care of the duplicate testing code and fixed all outstanding pep8 issues.

File: letsencrypt/cli.py [hunk 24]
```diff
@@ -185,15 +218,86 @@ def run(args, config, plugins):
         return "Configurator could not be determined"
 
     domains = _find_domains(args, installer)
+
```
Comment (jdkasten) [inline letsencrypt/cli.py]: I feel like this whole block of code can go in a function so that auth can simply call it

Comment (jdkasten) [inline letsencrypt/cli.py]: Maybe return treat_as_renewal?

**Comment by jdkasten** [comment]
Discussion with @pde, @schoen, we are going to bring this in and then immediately work to fix the auth issue.
